src/PolyType/ReflectionProvider/ReflectionTypeShape.cs

```diff
@@ -13,4 +16,19 @@ internal abstract class ReflectionTypeShape<T>(ReflectionTypeShapeProvider provi
     ITypeShapeProvider ITypeShape.Provider => provider;
     ICustomAttributeProvider? ITypeShape.AttributeProvider => typeof(T);
     object? ITypeShape.Invoke(ITypeShapeFunc func, object? state) => func.Invoke(this, state);
+
+    public Func<object>? GetAssociatedTypeFactory(Type relatedType)
+    {
+        if (relatedType.IsGenericTypeDefinition && this.Type.GenericTypeArguments.Length != relatedType.GetTypeInfo().GenericTypeParameters.Length)
+        {
+            throw new ArgumentException($"Related type arity ({relatedType.GenericTypeArguments.Length}) mismatch with original type ({this.Type.GenericTypeArguments.Length}).");
```

Comment:
I take it that the attribute annotations have no effect in the case of reflection? Can we document this?

Reply:
Sure. I expect you mean in shape-providers.md, where you've asked for some revisions already. I'll mix this one in.

Reply:
My only qualm with the current approach is that in reflection every type is an associated type. Ideally shapes coming from the two providers should be observationally equivalent.

Reply:
Ah right, because of the unreliability of querying all assemblies in the current domain?

Reply:
Right. Also, "all" assemblies isn't the same set of assemblies that an assembly was compiled with. So you could end up with a larger or smaller set of assemblies, leading to non-deterministic results.